### ISSUE TITLE
Documentation for macOS auth type 33

### ISSUE THREAD

**Issue by barneygale** [rfbproto/rfbproto#52]
I've partially reverse-engineered the auth type preferred by macOS VNC servers, ID 33. Implementation (in Python) available here:

https://github.com/barneygale/asyncvnc/blob/825447564e3af6b0d4a0996793f1ca7fb360c48f/asyncvnc.py#L480-L500

It's a bit like the SSH handshake. If the client doesn't know the server's public key, if it first queries it. Then it generates an AES key, and uses it to encrypt a 128-byte struct containing the username and password. It then encrypts the AES key using the server's public key, and sends both the RSA-encrypted AES key, and the AES-encrypted credentials, to the server. The server decrypted the AES key with its private key, then decrypts the username/password struct with the AES key.

There's a few bytes I'm not sure about, still.

**Comment by barneygale** [comment]
This an evolution of an earlier macOS auth type, ID 30, which is implemented in libvncserver:

https://github.com/LibVNC/libvncserver/blob/111db4bea124af27986c665dabd30c290168bb3d/src/libvncclient/rfbclient.c#L806-L924

**Comment by CendioOssman** [comment]
Nice work. You are welcome to submit a PR once you've figured out enough that things can be properly documented.

Security type 30 was recently documented in #43 based on the work done by @pauldumais for noVNC. @pdlan has also been working on figuring out some of the proprietary security types. Perhaps they can give some pointers here?

**Comment by Artoria2e5** [comment]
It feels like there could be some overlap with the "[RSA-AES Security Type](https://github.com/rfbproto/rfbproto/blob/master/rfbproto.rst#id37)" section, although the amount of reuse is probably very low. A less complete, but more struct-y description is in https://gitlab.com/Remmina/Remmina/-/wikis/Problems-and-tweaks/Apple-VNC-protocol-notes. A very rough draft follows:

### Apple security type 33

*all values are big-endian*

After this security type is selected, the client checks whether it has the server's pub key. If it does, it skips to the authentication step. If it does not, it sends a key request packet:

|length|type|value|description|
|----|----|----|----|
|4|U32|0x0a|length|
|2|U16|0x1000|version|
|4|U32|'RSA1'|algorithm|
|4|U32|0|reserved|

The server, on receiving a pubkey request, sends:

|length|type|description|
|----|----|----|
|4|U32|length|
|2|U16|version|
|4|U32|_server-key-length_|
|_server-key-length_|U8 array|key in DER format|
|1|U8|unknown|

The client begins authentication after obtaining the public key. It generates 16 random bytes as the AES key and uses it to encrypt a 128-byte structure of the username and password (defined in 7.2.12 Diffie-Hellman Authentication) in ECB mode.

The client encrypts the AES key using the server's RSA public key. The EME-PKCS1-v1_5 padding algorithm is used.

The client then sends the encrypted data:

|length|type|value|description|
|----|----|----|----|
|4|U32|0x018a|length|
|2|U16|0x1000|version|
|4|U32|'RSA1'|algorithm|
|2|U16|0x01|entry deliminator|
|128|U8 array||_encrypted-credentials_|
|2|U16|0x01|entry deliminator|
|_server-key-length_|U8 array|_encrypted-aeskey_|

After that server verifies if supplied credentials are correct and continues with the [SecurityResult](https://github.com/rfbproto/rfbproto/blob/master/rfbproto.rst#securityresult) message. etc etc

**Comment by Artoria2e5** [comment]
Huh, just noticed that the Remmina article mentions some "Secure Remote Password", but 33 is nowhere like SRP. https://support.apple.com/no-no/guide/remote-desktop/apdfe8e386b/mac mentions three separate auths:
* DH, that's type 30
* 2048-bit RSA host keys, that's type 33
* "SRP", which we probably haven't seen yet?? Is it type 36?

**Comment by barneygale** [comment]
I suspect SRP is type 36, yeah. It's been too long for me to remember the source of this suspicion, though 🙃.

Last I checked, the Mac VNC client (Cmd+K in Finder) prefers type 33, so it's probably not worth worrying about SRP?

**Comment by barneygale** [comment]
If fully implemented it should be secure. The Mac client sets *shared-flag* in the *ClientInit* packet to 0x1C, which causes the server to modify the format of the *ServerInit* packet it sends in response. Subsequent traffic appears to be encrypted -- I'd imagine it uses the already-established AES key, which should prevent MITMing.

**Comment by Artoria2e5** [comment]
It should; the linked Apple "apdfe8e386b" page lists 2048-bit RSA host keys for "Control and Observe", which I assume is basic VNC functionality. In fact, the page says everything in Apple RD can use some sort of encryption, way back to the DH days.

Good for them, bad for people trying to make compatible clients...

**Comment by cxzlw** [comment]
> I suspect SRP is type 36, yeah. It's been too long for me to remember the source of this suspicion, though 🙃.
> 
> Last I checked, the Mac VNC client (Cmd+K in Finder) prefers type 33, so it's probably not worth worrying about SRP?

Yes SRP is type 36. With IDA we can reverse engineering the `screensharingd` executable. And logs in the code help a lot (indicate purpose, naming and state).

Here are my conclusions: 

guest 8223（hex of 32 and 31）
vnc 2
kerberos 35
defaults? buf[1], buf[2]*(_WORD *)&buf[1] = 8478; 0x211e (0x21 0x1e) (type 33 30)
srp buf[3] = 36（v27 is 3 when SRP disabled, then buf[3] is replaced）
